cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -327,31 +329,38 @@ BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
     }
   }
 
+  ArrayData* output = out->array_data().get();
+  output->length = input.length;
+  output->SetNullCount(input.null_count);
+
+  // Set up validity bitmap
+  ARROW_ASSIGN_OR_RAISE(output->buffers[0],
+                        GetOrCopyNullBitmapBuffer(input, ctx->memory_pool()));
+
+  // Set up offset and data buffer
+  OffsetBuilder offset_builder(ctx->memory_pool());
+  RETURN_NOT_OK(offset_builder.Reserve(input.length + 1));
+  offset_builder.UnsafeAppend(0);  // offsets start at 0
   const int64_t sum_of_binary_view_sizes = util::SumOfBinaryViewSizes(
       input.GetValues<BinaryViewType::c_type>(1), input.length);
-
-  // TODO(GH-43573): A more efficient implementation that copies the validity
-  // bitmap all at once is possible, but would mean we don't delegate all the
-  // building logic to the ArrayBuilder implementation for the output type.
-  OutputBuilderType builder(options.to_type.GetSharedPtr(), ctx->memory_pool());
-  RETURN_NOT_OK(builder.Resize(input.length));
-  RETURN_NOT_OK(builder.ReserveData(sum_of_binary_view_sizes));
-  arrow::internal::ArraySpanInlineVisitor<I> visitor;
-  RETURN_NOT_OK(visitor.VisitStatus(
+  DataBuilder data_builder(ctx->memory_pool());
+  RETURN_NOT_OK(data_builder.Reserve(sum_of_binary_view_sizes));
+  RETURN_NOT_OK(VisitArraySpanInline<I>(
       input,
-      [&](std::string_view v) {
-        // Append valid string view
-        return builder.Append(v);
+      [&](std::string_view s) {
+        // for non-null value, append string view to buffer and calculate offset
+        data_builder.UnsafeAppend(reinterpret_cast<const uint8_t*>(s.data()),
+                                  static_cast<int64_t>(s.size()));
+        offset_builder.UnsafeAppend(static_cast<offset_type>(data_builder.length()));
+        return Status::OK();
       },
       [&]() {
-        // Append null
-        builder.UnsafeAppendNull();
+        // for null value, no need to update data buffer
+        offset_builder.UnsafeAppend(static_cast<offset_type>(data_builder.length()));
         return Status::OK();
       }));
```

Comment:
You don't need to return a Status at all, since errors cannot occur here. You should therefore be able to write:
```c++
  VisitArraySpanInline<I>(
      input,
      [&](std::string_view s) {
        // for non-null value, append string view to buffer and calculate offset
        data_builder.UnsafeAppend(reinterpret_cast<const uint8_t*>(s.data()),
                                  static_cast<int64_t>(s.size()));
        offset_builder.UnsafeAppend(static_cast<offset_type>(data_builder.length()));
      },
      [&]() {
        // for null value, no need to update data buffer
        offset_builder.UnsafeAppend(static_cast<offset_type>(data_builder.length()));
      }));
```

Reply:
done